### ISSUE TITLE
SOAP implementation details need to be checked

### ISSUE THREAD

**Issue by ardunn** [hackingmaterials/matminer#420]
Related to https://github.com/SINGROUP/dscribe/issues/29

@lauri-codes

The class can be found in `matminer.featurizers.structure` as `SOAP`.

**Comment by lauri-codes** [comment]
I'm now doing some minor changes to the Featurizer and I will open a pull request soon.

A couple of questions first:

1. What kind of SOAP default values do you want? Do you prefer speed or accuracy? The current ones (lmax=2, nmax=4, rcut=3.0, sigma=1.0) are fast but not so accurate in reproducing the environment.
2. The Featurizer currently only provides the average SOAP output (averaged over all sites in the structure). The full SOAP output that contains a vector for each atom would be more general, but I don't know how this fits the matminer-thinking. Do you only support a single vector per structure, or is it possible to output a 2D-array for each structure? Another possibility would be to use e.g. the [REMatch kernel](https://singroup.github.io/dscribe/tutorials/kernels.html#rematch-kernel) that still provides a single vector/structure.

**Comment by computron** [comment]
Hi Lauri,

Thanks for getting involved! Some responses to your questions

1. The general policy we have followed so far is:
- The constructor doesn't specify defaults for the arguments when there are multiple "good" answers (i.e., we are not 80-90% certain that the user would want to use a single default).
- Instead, we add a method to the featurizer called "from_preset()", which takes in a String argument. Thus, your from_preset() might have possible strings called "fast" or "accurate" that initialize the featurizer with the user-desired behavior
- If you search for "from_preset" in the codebase, you should see multiple places where this was done.

2. The featurize() method is intended to return only a single array (not a 2D array). There are a few different ways to address your concerns:
- The featurizers.site package contains featurizers that are intended to return an array per site (i.e., per atom), not an array per structure. Thus, anything where you are featurizing sites could go in that package.
- The featurizers.structure package should return an array per structure. You could also create an additional featurizer here that does the individual site descriptors and transforms to a 1D array using whatever algorithm you prefer. Note that if there are multiple algorithms to choose from, this can be a parameter in the constructor and might also be a candidate for "from_preset()" .

3. Note that the featurizers.utils.stats.py module contains functions for reducing the dimensions of things - e.g., turning a 1D array into a number (e.g., by taking the mean) or turning a matrix into a vector (by taking the eigenvalues). While I am not sure the REMatch kernel should go in here, I just wanted to alert you to this module. Usually, the featurizers themselves will let you pick which stats function you want to use to reduce the dimension of the output.

Let me know if you have any further questions, and thanks again!

**Comment by ardunn** [comment]
Hey @lauri-codes 
I think we could have several presets, like @computron said. In general, slow but accurate > fast and inaccurate.

If seems like we should probably accomplish 2 tasks here:

1. put a site version in `matminer.featurizers.site`
2. get an average structure SOAP either by
    a. use the SiteStatsFingerprint class in matminer, or...
    b. using the native dscribe implementation with `average` specified

Whether we choose 2a or 2b will depend on speed. Matminer uses multiprocessing in python which speeds up over single processes but are slow compared to c++ multithreading obviously. Does dscribe have methods for parallelizing on a single node?

**Comment by lauri-codes** [comment]
1. I will put SOAP into matminer.featurizers.site where it naturally belongs. 
3. DScribe doesn't provide parallelization on the C++ side, only through python with joblib. So it definitely makes sense to simply use SiteStatsFingerprint since you already have it. Would it then make sense to completely remove the current SOAP-class that is in matminer.featurizers.structure?

**Comment by lauri-codes** [comment]
There is now a pull request which puts SOAP into featurizers.site. When you have the time please review it and make sure the changes are appropriate.

**Comment by ardunn** [comment]
Hey @lauri-codes, do you happen to have any code containing matminer's implementaiton of SOAP (or even just dscribe) which demonstrates the features as useful for prediction? I have tried them on a few thermodynamic and vibrational properties without much luck.

**Comment by ardunn** [comment]
@lauri-codes In particular, SOAP averaged over entire structures - similar to the experiment you guys report on the dscribe paper?

**Comment by lauri-codes** [comment]
Hi @ardunn,

Here is a small script that demonstrates the kind of learning curve that we also observe in the paper:
```python
import random
import math
import ase
import ase.io
import numpy as np
import matplotlib.pyplot as mpl
from dscribe.descriptors import SOAP
from sklearn.kernel_ridge import KernelRidge
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_absolute_error

# Load dataset
samples = ase.io.read("./formation_energy.extxyz", index=":")
random.seed(42)
random.shuffle(samples)
samples = samples[0:16384]

samples = ase.io.read("./formation_energy_16384_random.extxyz", index=":")
Y = np.array([x.info["formation_energy_ev_per_atom"] for x in samples])
atomic_numbers = set()
for system in samples:
    i_atomic_numbers = set(system.get_atomic_numbers())
    atomic_numbers = atomic_numbers.union(i_atomic_numbers)

# Setup the descriptor with some defaults
descriptor = SOAP(
    species=atomic_numbers,
    rcut=6,
    nmax=8,
    lmax=8,
    sigma=0.4,
    periodic=True,
    crossover=True,
    average=True,
    sparse=True,
)

# Precalculate descriptors
X = descriptor.create(samples, n_jobs=4)

# Loop over different dataset sizes:
sizes = [2048, 4096, 8192, 16384]
alpha = 1e-2
gamma = 2**(-8)
maes = []
for size in sizes:

    # Setup model
    krr = KernelRidge(kernel="rbf", alpha=alpha, gamma=gamma)

    # Split dataset into training and test
    x_train, x_test, y_train, y_test = train_test_split(
        X,
        Y,
        test_size=math.ceil(0.2*size),
        train_size=math.floor(0.8*size),
        random_state=42,
        shuffle=True
    )

    # Train
    krr.fit(x_train, y_train)

    # Test
    y_predicted = krr.predict(x_test)
    mae = mean_absolute_error(y_test, y_predicted)
    maes.append(mae)

mpl.plot([math.floor(0.8*x) for x in sizes], maes)
mpl.xlabel("Training set size")
mpl.ylabel("Mean absolute error (eV/atom)")
mpl.title("Learning curve for formation energy prediction with SOAP averaged over atoms")
mpl.show()
```

This can be run in 5 minutes on a local machine. The learning setup is very simple but should demonstrate the proper learning curve. The dataset you can download from [FigShare](https://figshare.com/collections/Supplementary_datasets_for_DScribe_Library_of_Descriptors_for_Machine_Learning_in_Materials_Science_/4607783). If you run that script, you should be able to get a plot like this:

![formation_energy_soap_learning_curve](https://user-images.githubusercontent.com/6513318/88377496-ff9e0500-cda7-11ea-83bc-281314660336.png)

This was run today on my linux laptop with the following pip setup:
```sh
Package                       Version     
----------------------------- ------------
alabaster                     0.7.12      
ase                           3.19.0      
astroid                       2.3.3       
attrs                         19.3.0      
Babel                         2.8.0       
certifi                       2019.11.28  
chardet                       3.0.4       
coverage                      5.0.3       
coveralls                     1.10.0      
cycler                        0.10.0      
Cython                        0.29.14     
decorator                     4.4.1       
docopt                        0.6.2       
docutils                      0.16        
dscribe                       0.3.5       
future                        0.18.2      
idna                          2.8         
imagesize                     1.2.0       
importlib-metadata            1.6.0       
isort                         4.3.21      
Jinja2                        2.11.1      
joblib                        0.14.1      
kiwisolver                    1.1.0       
latexcodec                    2.0.0       
lazy-object-proxy             1.4.3       
MarkupSafe                    1.1.1       
matplotlib                    3.1.2       
mccabe                        0.6.1       
monty                         3.0.2       
more-itertools                8.2.0       
mpmath                        1.1.0       
networkx                      2.4         
numpy                         1.18.1      
oset                          0.1.3       
packaging                     20.1        
palettable                    3.3.0       
pandas                        1.0.1       
pip                           20.0.2      
pluggy                        0.13.1      
py                            1.8.1       
pybind11                      2.4.3       
pybtex                        0.22.2      
pybtex-docutils               0.2.2       
PyDispatcher                  2.0.5       
Pygments                      2.5.2       
pylint                        2.4.4       
pymatgen                      2020.1.28   
pyparsing                     2.4.6       
pytest                        5.4.1       
python-dateutil               2.8.1       
pytz                          2019.3      
PyYAML                        5.3         
requests                      2.22.0      
ruamel.yaml                   0.16.7      
ruamel.yaml.clib              0.2.0       
scikit-learn                  0.22.1      
scipy                         1.4.1       
setuptools                    46.1.3      
six                           1.13.0      
sklearn                       0.0         
snowballstemmer               2.0.0       
soaplite                      1.0.3       
spglib                        1.14.1.post0
Sphinx                        3.1.2       
sphinx-rtd-theme              0.5.0       
sphinxcontrib-applehelp       1.0.1       
sphinxcontrib-bibtex          1.0.0       
sphinxcontrib-devhelp         1.0.1       
sphinxcontrib-htmlhelp        1.0.2       
sphinxcontrib-jsmath          1.0.1       
sphinxcontrib-qthelp          1.0.2       
sphinxcontrib-serializinghtml 1.1.3       
sympy                         1.5.1       
tabulate                      0.8.6       
typed-ast                     1.4.0       
urllib3                       1.25.8      
wcwidth                       0.1.9       
wheel                         0.33.6      
wrapt                         1.11.2      
zipp                          3.1.0
```

**Comment by ardunn** [comment]
Awesome thank you @lauri-codes
